lib/compute_visible_services.ts

```diff
@@ -0,0 +1,127 @@
+import * as winston from "winston";
+import * as documentDbUtils from "../lib/utils/documentdb";
+
+import { IContext } from "azure-function-express";
+import { DocumentClient } from "documentdb";
+import { isLeft } from "fp-ts/lib/Either";
+import { isNone, isSome } from "fp-ts/lib/Option";
+import { pick } from "italia-ts-commons/lib/types";
+import { RetrievedService } from "./models/service";
+import {
+  VISIBLE_SERVICE_COLLECTION_NAME,
+  VisibleServiceModel
+} from "./models/visible_service";
+import { getRequiredStringEnv } from "./utils/env";
+import { PermanentError } from "./utils/errors";
+import { configureAzureContextTransport } from "./utils/logging";
+
+// Whether we're in a production environment
+const isProduction = process.env.NODE_ENV === "production";
+
+const cosmosDbUri = getRequiredStringEnv("CUSTOMCONNSTR_COSMOSDB_URI");
+const cosmosDbKey = getRequiredStringEnv("CUSTOMCONNSTR_COSMOSDB_KEY");
+const cosmosDbName = getRequiredStringEnv("COSMOSDB_NAME");
+
+const documentDbDatabaseUrl = documentDbUtils.getDatabaseUri(cosmosDbName);
+
+const visibleServicesCollectionUrl = documentDbUtils.getCollectionUri(
+  documentDbDatabaseUrl,
+  VISIBLE_SERVICE_COLLECTION_NAME
+);
+const documentClient = new DocumentClient(cosmosDbUri, {
+  masterKey: cosmosDbKey
+});
+
+const visibleServiceModel = new VisibleServiceModel(
+  documentClient,
+  visibleServicesCollectionUrl
+);
+
+/**
+ * A function to cache the list of visible (active) services.
+ * Reacts to CosmosDB changes (change feed) in the services collection.
+ */
+export async function index(
+  context: IContext<{}>,
+  documents: ReadonlyArray<RetrievedService>
+): Promise<void> {
+  const logLevel = isProduction ? "info" : "debug";
+  configureAzureContextTransport(context, winston, logLevel);
+
+  try {
+    documents.forEach(async createdOrUpdatedService => {
+      winston.debug(
+        "ComputeVisibleService: %s",
+        JSON.stringify(createdOrUpdatedService)
+      );
+      const errorOrMaybeVisibleService = await visibleServiceModel.find(
+        createdOrUpdatedService.id,
+        createdOrUpdatedService.serviceId
+      );
+      if (isLeft(errorOrMaybeVisibleService)) {
+        winston.debug(
+          "ComputeVisibleService Query Error: %s",
+          JSON.stringify(errorOrMaybeVisibleService.value)
+        );
+        return PermanentError("QueryError querying for visible service");
```

Comment:
that's right, but what can we do in case of transient error here ? (how to schedule a retry ? : )

Reply:
shouldn't throwing an exception make the event being reprocessed?

Reply:
yes, if you trigger an exception here the events will be rescheduled using the default functions configurations (a burst of fast re-processing). 

this function is triggered by a cosmosdb change feed event which cannot be re-scheduled (if not re-saving the service inside the services collection). we may schedule "inner" retries but this will stretch the function runtime.